operator/cmd/mesh/uninstall.go

```diff
@@ -0,0 +1,200 @@
+		"The namespace of Istio Control Plane.")
+	cmd.PersistentFlags().StringVarP(&args.filename, "filename", "f", "",
+		"The filename of the IstioOperator CR.")
+	cmd.PersistentFlags().StringVarP(&args.manifestsPath, "manifests", "d", "", ManifestsFlagHelpStr)
+	cmd.PersistentFlags().StringArrayVarP(&args.set, "set", "s", nil, setFlagHelpStr)
+}
+
+func UninstallCmd(logOpts *log.Options) *cobra.Command {
+	rootArgs := &rootArgs{}
+	uiArgs := &uninstallArgs{}
+	uicmd := &cobra.Command{
+		Use:   "uninstall --revision foo or uninstall --filename iop.yaml",
+		Short: "uninstall the control plane by revision",
+		Long:  "The uninstall command uninstall the control plane by revision or IstioOperator CR",
+		Example: `  # uninstall by revision label foo
+  istioctl uninstall --revision foo
+
+  # uninstall by iop file
+  istioctl uninstall -f iop.yaml
+  
+  # uninstall all istio resources
+  istioctl uninstall --purge
+`,
+		Args: func(cmd *cobra.Command, args []string) error {
+			if uiArgs.revision == "" && uiArgs.filename == "" && !uiArgs.purge {
+				return fmt.Errorf("at least one of the --revision, --filename or --purge flags must be set")
+			}
+			if len(args) > 0 {
+				return fmt.Errorf("istioctl uninstall does not take arguments")
+			}
+			return nil
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return uninstall(cmd, rootArgs, uiArgs, logOpts)
+		}}
+	addUninstallFlags(uicmd, uiArgs)
+	return uicmd
+}
+
+// uninstall uninstalls control plane by either pruning by target revision or deleting specified manifests.
+func uninstall(cmd *cobra.Command, rootArgs *rootArgs, uiArgs *uninstallArgs, logOpts *log.Options) error {
+	l := clog.NewConsoleLogger(cmd.OutOrStdout(), cmd.ErrOrStderr(), installerScope)
+	if err := configLogs(logOpts); err != nil {
+		return fmt.Errorf("could not configure logs: %s", err)
+	}
+	client, err := k8sClient(uiArgs.kubeConfigPath, uiArgs.context)
+	if err != nil {
+		return err
+	}
+
+	cache.FlushObjectCaches()
+	opts := &helmreconciler.Options{DryRun: rootArgs.dryRun, Log: l, ProgressLog: progress.NewLog()}
+	h, err := helmreconciler.NewHelmReconciler(client.Controller(), client.RESTConfig(), nil, opts)
+	if err != nil {
+		return fmt.Errorf("failed to create reconciler: %v", err)
+	}
+
+	// If only revision flag is set, we would prune resources by the revision label.
+	// Otherwise we would merge the revision flag and the filename flag and delete resources by generated manifests.
+	if uiArgs.filename == "" {
+		emptyiops := &v1alpha1.IstioOperatorSpec{Profile: "empty", Revision: uiArgs.revision}
+		iop, err := translate.IOPStoIOP(emptyiops, "empty", iopv1alpha1.Namespace(emptyiops))
+		if err != nil {
+			return err
+		}
+		h, err := helmreconciler.NewHelmReconciler(client.Controller(), client.RESTConfig(), iop, opts)
+		if err != nil {
+			return fmt.Errorf("failed to create reconciler: %v", err)
+		}
+		objectsList, tp, err := h.GetPrunedResources(uiArgs.revision, uiArgs.purge)
```

Comment:
I first did `uninstall --revision default`, then `uninstall --purge`.

I had a lot of resources left around after the uninstall, including the istio-ingressgateway.  When I did the `--purge` it complained "Error: failed to do preuninstall check: failed to get proxy infos: unable to find any Pilot instances"

Does `--purge` need an istiod instance?  It should be able to remove without one.  Also, the message should be "unable to find any istiod instances", not "Pilot instances".

If you `--purge` more than once logging errors will appear.  To get rid of them we can consider adding "install" to the list scope log levels in _root.go_ `defaultLogOptions()`.

Reply:
for the case of `uninstall --revision default`, if those resources are not removed most likely they don't have the targeted revision label. And I think the ingressgateway is actually upgraded in place to the newer version? There is no revision for ingressgateway and there are no multiple ingressgateways with different revisions exist.

Ok for the purge case then it should be just a warning and not an error.